bn_mp_pack.c

```diff
@@ -36,16 +36,17 @@ mp_err mp_pack(void *rop, size_t *countp, mp_order order, size_t size,
 
    odd_nails = (nails % 8u);
    odd_nail_mask = 0xff;
-   for (i = 0; i < odd_nails; ++i) {
+   for (i = 0u; i < odd_nails; ++i) {
       odd_nail_mask ^= (unsigned char)(1u << (7u - i));
    }
    nail_bytes = nails / 8u;
 
    bits = (size_t)mp_count_bits(&t);
    count = (bits / ((size * 8u) - nails)) + (((bits % ((size * 8u) - nails)) != 0u) ? 1u : 0u);
```

Comment:
Use mp_pack_size/count here?

Reply:
> Use mp_pack_size/count here?

Yes, of course.

So you are OK with what I did?

Reply:
Yes, but we still have to decide bytes or units.

Right now countp in units, maxsize in bytes, pack_size in bytes? The same applies to mp_unpack.

Reply:
I shuffled it a bit around, probably more to what you wanted in the first place.
I hope ;-)

Argh, documentation not yet updated!

> Yes, but we still have to decide bytes or units.

Do you want to name them (`enum`)?
Otherwise the types can stay as they are now.

Reply:
Hmm I don't like it like this. We have count for unpack which is used to pass in the buffer size and for pack we have both maxsize and countp. Either we also change countp/count to bytes or we change maxsize to units. Mixing the units is not good as we have it now.

Reply:
(I hope I'm up-to-date now with the rebasing)

> Hmm I don't like it like this.
Yeah, it's a bit of a mess, admitted.

> Either we also change countp/count to bytes or we change maxsize to units.

`maxsize` (I think it's `maxlen` elsewhere but I always mix them up) is in all other cases meant to measure the given buffer-size in bytes, a `size_t` type.

Let me restrict myself to `mp_pack` for now:
`countp` does, as not only the name suggests, some (ac)counting. The size in bytes behind `countp` depends on a second variable (it formerly returned `count` which were units computed from bitsize and nails) and can be different from the amount of actually used up bytes. so I would say that the expectation is until now that `countp` returns units.
You won't have a port to get the amount of written bytes if you use `countp` for units.

On the other side: we have an external function to compute `countp` so the information that`countp` was once needed for is now available externally and in advance, `countp` is now free for other uses, e.g.: returning the number bytes written.
It changes the API but it is not much work to write a little macro as a wrapper if so desired.

So for `mp_pack` I'd propose:
 - `countp` (maybe change the name) are the bytes written
 - `maxsize` is the size in bytes of the buffer given
 - `count` as computed by `mp_pack_count` returns the number of units you get with the combination of bit-size and nails given to `mp_pack_count`. That means that there are no accounts of units send in or out of `mp_pack`.

And for `mp_unpack`
 - `count` is not the number of bytes but the number of units, the actual byte-size of the input is only of internal interest but can be computed from `count` and `nails` if so desired although I would just measure the input.

Reply:
pack/unpack should also be symmetric. This means count should be bytes/unit for both. However we should also consider that changing from units/bytes will introduce errors for user who do not expect that change when they move to the new function.

Reply:
My suggestion would therefore be to call it mp_pack_count, count counts unit, maxcount etc. This also makes sense. The user is not operating on byte arrays but on arrays of unit size (like calloc for example). The only downside is that its different from the sbin/ubin functions which operate on bytes.